Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+    mutating func writabilityChanged(writable: Bool) -> Action {
+        self.isChannelWritable = writable
+
+        switch self.state {
+        case .initialized,
+             .finished,
+             .failed:
+            return .wait
+
+        case .running(.verifyRequest, _), .running(.endSent, _):
+            return .wait
+
+        case .running(.streaming(let expectedBody, let sentBodyBytes, producer: .paused), let responseState):
+            if writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBody,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .producing
+                )
+
+                self.state = .running(requestState, responseState)
+                return .resumeRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, producer: .producing), let responseState):
+            if !writable {
+                let requestState: RequestState = .streaming(
+                    expectedBodyLength: expectedBodyLength,
+                    sentBodyBytes: sentBodyBytes,
+                    producer: .paused
+                )
+                self.state = .running(requestState, responseState)
+                return .pauseRequestBodyStream
+            } else {
+                // no state change needed
+                return .wait
+            }
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        return .read
+    }
+
+    mutating func errorHappened(_ error: Error) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("After the state machine has been initialized, start must be called immidiatly. Thus this state is unreachable")
+        case .running:
+            self.state = .failed(error)
+            return .failRequest(error, closeStream: true)
+        case .finished, .failed:
+            preconditionFailure("If the request is finished or failed, we expect the connection state machine to remove the request immidiatly from its state. Thus this state is unreachable.")
+        }
+    }
+
+    mutating func start() -> Action {
+        guard case .initialized = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .running(.verifyRequest, .initialized)
+        return .verifyRequest
+    }
+
+    mutating func requestVerified(_ head: HTTPRequestHead) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        guard self.isChannelWritable else {
+            preconditionFailure("Unimplemented. Wait with starting the request here!")
+        }
+
+        if let value = head.headers.first(name: "content-length"), let length = Int(value), length > 0 {
```

Comment:
This code should deal with the possibility the user has provided multiple, inconsistent transport header fields (i.e. multiple content-lengths _and_ a transfer-encoding: chunked).

Reply:
My idea here was that the state machine triggers the action: `return .verifyRequest`. It is the callers task to then call `HTTPHeaders.validate` that is currently in `RequestValidation.swift`. This should ensure that the request head and headers are valid. We could of course move the request validation directly into the state machine. (I didn't be cause it may change the http request, which I would consider a side effect).

Reply:
For reference: https://github.com/swift-server/async-http-client/blob/main/Sources/AsyncHTTPClient/RequestValidation.swift